File: Python/wordPattern.py

pattern = "abba" 
s = "dog constructor constructor dog"

def wordPattern(s, pattern):

    dictionary = {}
    
    pattern_map = set()
    mapStr_s = s.split(" ")

    if len(mapStr_s) != len(pattern):
        return False
    
    for char_s, char_pattern in zip(mapStr_s, pattern):
      
        if char_s in dictionary:
            if dictionary[char_s] != char_pattern:
                return False
        else:
            if char_pattern in pattern_map:
                return False

            dictionary[char_s] = char_pattern
            pattern_map.add(char_pattern)
  
    return True 


print(wordPattern(s, pattern))    

File: JavaScript/wordPattern.js


let pattern = "abba" 
let s = "dog constructor constructor dog"

var wordPattern = function (s, pattern) {
    let mapStr_s = s.split(" ");
    let dict = new Map();
    let mapStr_t = []
    let x = 0
    console.log(mapStr_s)

     if (pattern.length != mapStr_s.length)  return false  
        
       for (let str_s of mapStr_s) {

        if (dict.has(str_s)) {
            console.log(str_s)
            if (dict.get(str_s) != pattern[x]) return false
        }
        else if (mapStr_t.includes(pattern[x])) {
            return false
        }
        else {
            dict.set(str_s, pattern[x]);
            mapStr_t.push(pattern[x])           
        }
        x++
        console.log(dict)
    }

    return true

};

console.log(wordPattern(s, pattern))